### ISSUE TITLE
EAS Production build not working for iOS

### ISSUE THREAD

**Issue by jbtheard** [customerio/customerio-expo-plugin#64]
My EAS production build for iOS fails with that error:

```
[RUN_FASTLANE] $ xcodebuild -showBuildSettings -workspace ./MyApp.xcworkspace -scheme MyApp -configuration Release
[RUN_FASTLANE] 2023-04-24 15:34:41.639 xcodebuild[4740:10663925] DVTCoreDeviceEnabledState: DVTCoreDeviceEnabledState_Disabled set via user default (DVTEnableCoreDevice=disabled)
[RUN_FASTLANE] 2023-04-24 15:34:43.087 xcodebuild[4740:10663930]  DVTPortal: Service '<DVTPortalViewDeveloperService: 0x60000206f100; action='viewDeveloper'>' encountered an unexpected result code from the portal ('1100')
[RUN_FASTLANE] 2023-04-24 15:34:43.087 xcodebuild[4740:10663930]  DVTPortal: Error:
[RUN_FASTLANE] Error Domain=DVTPortalServiceErrorDomain Code=1100 "Your session has expired. Please log in." UserInfo={payload={
[RUN_FASTLANE]     creationTimestamp = "2023-04-24T14:34:43Z";
[RUN_FASTLANE]     httpCode = 200;
[RUN_FASTLANE]     protocolVersion = QH65B2;
[RUN_FASTLANE]     requestUrl = "https://developerservices2.apple.com/services/QH65B2/viewDeveloper.action";
[RUN_FASTLANE]     responseId = [a uuid];
[RUN_FASTLANE]     resultCode = 1100;
[RUN_FASTLANE]     resultString = "Your session has expired. Please log in.";
[RUN_FASTLANE]     userLocale = "en_US";
[RUN_FASTLANE]     userString = "Your session has expired. Please log in.";
[RUN_FASTLANE] }, NSLocalizedDescription=Your session has expired. Please log in.}
```

When I build for simulator, it does build properly though.

The only difference with previous successful build is the addition of customer.io packages.

Please note I use expo-notification as a primary way to handle notifications; could it be a conflict with that library ?

**Comment by jbtheard** [comment]
Yes, rich push is indeed enabled, and I assume the "host" app code signing has been automatically setup by EAS.

To be fair, I was a bit surprised to see doc related to Xcode in Expo doc (see [here](https://github.com/customerio/customerio-expo-plugin/issues/65)).

How can I configure this new target without having to touch the Xcode project (which isn't necessary with expo)

**Comment by levibostian** [comment]
Thanks for that feedback on the Expo plugin. 

When enabling rich push via the Customer.io Expo plugin, a new Xcode target must be created and code signing must be created for that new Xcode target. This is something required by Apple that we cannot get around. 

However, I do agree with you that it would be ideal if you could resolve this requirement without having to interact with Xcode. 

Do you remember how you setup iOS code signing for your Expo app? When I say *iOS code signing* I am referring to [certificates and provisioning profiles](https://docs.expo.dev/app-signing/app-credentials/#ios). It would be nice to hear how developers have their code signing setup so we can try to make our Expo plugin compatible with that existing setup. That's why I ask if you remember how you set yours up. Did you [use the EAS CLI to setup EAS](https://docs.expo.dev/app-signing/managed-credentials/) to manage your credentials? Or did you follow some other documentation?

**Comment by jbtheard** [comment]
Also, you're right, removing useRichPush did allow the build to go through.

**Comment by jbtheard** [comment]
On your last point, how did you create the 2nd bundle? Via eas-cli or via their UI? I tried yesterday via UI but afaik that meant creating a new bundle in App Store Connect as well. Usually expo/eas handle everything so the less I do in Apple the happier I am.I also though about using wildcards when declaring key in App Store Connect but not sure if that would work…iThumbed, sorry for typos.On 25 Apr 2023, at 16:17, Levi Bostian ***@***.***> wrote:﻿
In regards to this comment, I am talking about iOS code signing files such as certificates and provisioning profiles that are used for signing your iOS app. Not files that are working with push.

adding this useRichPush options essentially adds a new target to the project with a new bundle id (previous.bundle.richpush).

Correct. The Customer.io expo plugin does more then just adding a new target but that is one part of it, yes.

Is that really necessary ?

Yes. All of the push notification features that Customer.io provides such as images, deep links, push metrics tracking all require that a new target gets created in the Xcode project. This is a requirement by Apple and something we cannot get around.
The goal of our Expo plugin is to make this Apple requirement convenient for you to setup. That is what I hope what conversations such as this one accomplishes.

Because I can't see a way to have EAS automatically handle keys for 2 different targets....

I am not an EAS expert, but I believe that this screenshot shows where in EAS you handle iOS code signing files. Notice in the screenshot that there are 2 separate targets (aka: bundle identifiers) where each entry contains a certificate and provisioning profile used to sign your app.

I would expect that this is where you setup your EAS server for X.richpush target code signing. After the EAS credentials are setup for this new X.richpush bundle identifier, I would expect your EAS builds to succeed.

**Comment by levibostian** [comment]
> On your last point, how did you create the 2nd bundle? Via eas-cli or via their UI? 

Via the EAS UI. I have not used the EAS CLI very much but I imagine that you could use `eas credentials`? I am not sure on the CLI part, sorry. 

> I tried yesterday via UI but afaik that meant creating a new bundle in App Store Connect as well. 

You will need [register a new app id](https://developer.apple.com/help/account/manage-identifiers/register-an-app-id) with your Apple Developer account, but I do not believe that you need to create a new app in App Store Connect. You might, I am not sure. It is possible to have a bundle ID registered to your Apple Developer account (developer.apple.com) but not be an app in App Store Connect (appstoreconnect.apple.com). 

I suggest you start with [registering a new app id](https://developer.apple.com/help/account/manage-identifiers/register-an-app-id) with your Apple Developer account (because that is required) and later on create an app in App Store Connect if Apple gives you errors when trying to publish an update to the App Store. 

> Usually expo/eas handle everything so the less I do in Apple the happier I am.I also though about using wildcards when declaring key in App Store Connect but not sure if that would work…

Wildcards for bundle IDs are only available for development code signing files. If you want to make adhoc builds or release your app to the App Store, wildcards are not allowed. 

---

I know that you prefer to stay far away from native Android and iOS concepts. Just in case you were curious about why the Expo plugin adds a new target, why you need to register a new app id, etc. [you can read about Notification Service Extensions](https://developer.apple.com/documentation/usernotifications/modifying_content_in_newly_delivered_notifications) which is what I am talking about. It's a requirement by Apple if you want to use any of the more advanced features in your push notifications. This is not specific to Customer.io, but all push services.

**Comment by deantes** [comment]
In order for EAS to pick up the new target in a managed environment you can add an expo extension. Documentation found here https://docs.expo.dev/build-reference/app-extensions/

` build: {
          experimental: {
            ios: {
              appExtensions: [
                {
                  targetName: 'NotificationService',
                  bundleIdentifier: `${appIdentifier}.richpush`,
                },
              ],
            },
          },
        },
        `
With this EAS will generate new provisioning certificates for the richpush target.

**Comment by jbtheard** [comment]
I confirm this fixes the build and this issue can be closed (but perhaps it should first be reported in the Expo doc)

**Comment by levibostian** [comment]
Thank you so much, @deantes for the suggestion and @jbtheard for the confirmation. 

I have submitted an update to our docs to include this suggestion. 

Great solution! 🎊
